lib/streamlit/elements/widgets/audio_input.py

```diff
@@ -0,0 +1,248 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from dataclasses import dataclass
+from textwrap import dedent
+from typing import TYPE_CHECKING, Union, cast
+
+from typing_extensions import TypeAlias
+
+from streamlit.elements.form_utils import current_form_id
+from streamlit.elements.lib.policies import (
+    check_widget_policies,
+    maybe_raise_label_warnings,
+)
+from streamlit.elements.lib.utils import (
+    Key,
+    LabelVisibility,
+    compute_and_register_element_id,
+    get_label_visibility_proto_value,
+    to_key,
+)
+from streamlit.elements.widgets.file_uploader import _get_upload_files
+from streamlit.proto.AudioInput_pb2 import AudioInput as AudioInputProto
+from streamlit.proto.Common_pb2 import FileUploaderState as FileUploaderStateProto
+from streamlit.proto.Common_pb2 import UploadedFileInfo as UploadedFileInfoProto
+from streamlit.runtime.metrics_util import gather_metrics
+from streamlit.runtime.scriptrunner import ScriptRunContext, get_script_run_ctx
+from streamlit.runtime.state import (
+    WidgetArgs,
+    WidgetCallback,
+    WidgetKwargs,
+    register_widget,
+)
+from streamlit.runtime.uploaded_file_manager import DeletedFile, UploadedFile
+
+if TYPE_CHECKING:
+    from streamlit.delta_generator import DeltaGenerator
+
+if TYPE_CHECKING:
+    from streamlit.delta_generator import DeltaGenerator
+
+SomeUploadedAudioFile: TypeAlias = Union[UploadedFile, DeletedFile, None]
+
+
+@dataclass
+class AudioInputSerde:
+    def serialize(
+        self,
+        audio_file: SomeUploadedAudioFile,
+    ) -> FileUploaderStateProto:
+        state_proto = FileUploaderStateProto()
+
+        if audio_file is None or isinstance(audio_file, DeletedFile):
+            return state_proto
+
+        file_info: UploadedFileInfoProto = state_proto.uploaded_file_info.add()
+        file_info.file_id = audio_file.file_id
+        file_info.name = audio_file.name
+        file_info.size = audio_file.size
+        file_info.file_urls.CopyFrom(audio_file._file_urls)
+
+        return state_proto
+
+    def deserialize(
+        self, ui_value: FileUploaderStateProto | None, widget_id: str
+    ) -> SomeUploadedAudioFile:
+        upload_files = _get_upload_files(ui_value)
+        if len(upload_files) == 0:
+            return_value = None
+        else:
+            return_value = upload_files[0]
+        return return_value
+
+
+class AudioInputMixin:
+    @gather_metrics("audio_input")
+    def audio_input(
+        self,
+        label: str,
+        key: Key | None = None,
+        on_change: WidgetCallback | None = None,
+        args: WidgetArgs | None = None,
+        kwargs: WidgetKwargs | None = None,
+        *,  # keyword-only arguments:
+        disabled: bool = False,
+        label_visibility: LabelVisibility = "visible",
+    ) -> UploadedFile | None:
```

Comment:
we unfortunately do not handle this consistently across widgets, but I want to highlight that using a Streamlit-defined type might not be the best UX because the developer sees the type, but has to click on it to get what it really means and we discourage developers to import it from here because we don't give any guarantees about our internal package structure.
I believe we should move all of our type definitions into a module `streamlit.types` or similar and make that an official part of our API (guarantee the existence) at some point. Until then, it might be nicer to duplicate the definition for the publicly exposed API here and write `Union[UploadedFile, DeletedFile, None]` (although also here two Streamlit-types are part of, so probably not worth it 😅)

No immediate action item for this PR, but wanted to share one inconsistency in our codebase right now so that we can all think about it 🙂

Reply:
@raethlein I definitely agree that types as part of Streamlit API is a great idea! 
We need to think carefully about exact types (especially for file-uploader, data-editor like widgets). But on the strategic level, it is something we should do at some point!